[css-anchor-position-1] Anchored element should be in the same or higher top layer than the anchor

anttijk opened the pull request:
#### c316ddee56b568bdfef3ffaa93ae589464da0a13
<pre>
[css-anchor-position-1] Anchored element should be in the same or higher top layer than the anchor
<a href="https://bugs.webkit.org/show_bug.cgi?id=295516">https://bugs.webkit.org/show_bug.cgi?id=295516</a>
<a href="https://rdar.apple.com/155216924">rdar://155216924</a>

Reviewed by Tim Nguyen.

<a href="https://drafts.csswg.org/css-anchor-position-1/#acceptable-anchor-element">https://drafts.csswg.org/css-anchor-position-1/#acceptable-anchor-element</a>

possible anchor is laid out strictly before positioned el, aka one of the following is true:
- positioned el is in a higher top layer than possible anchor
- Both elements are in the same top layer..

* LayoutTests/TestExpectations:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::computeTopLayerStatus):

Canonical link: <a href="https://commits.webkit.org/297070@main">https://commits.webkit.org/297070@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a75819542655deec78a8214b9349168aa982652e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110445 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30104 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20537 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116470 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60699 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112408 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30783 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38691 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83990 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113393 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24569 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99450 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64431 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23932 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17589 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60266 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93953 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17645 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119261 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37485 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27816 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92957 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37859 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95718 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92780 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23639 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37778 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15515 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33431 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37380 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42851 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37042 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40382 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38750 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a75819542655deec78a8214b9349168aa982652e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47657

nt1m commented on Source/WebCore/style/AnchorPositionEvaluator.cpp:
```diff
@@ -854,6 +854,44 @@ static CheckedPtr<const Element> anchorScopeForAnchorName(const RenderBoxModelOb
+    if (!anchored.document().hasTopLayerElement())
+        return TopLayerStatus::Same;
+
+    auto topLayerRoot = [&](auto& renderer) -> const RenderLayerModelObject* {
```
I would call this `topLayerAncestor` / `rootInTopLayer` because `topLayerRoot` suggests it's the root of all the top layer elements (which is the RenderView or wherever we append top layer stuff nowadays)

nt1m commented on Source/WebCore/style/AnchorPositionEvaluator.cpp:
```diff
@@ -854,6 +854,44 @@ static CheckedPtr<const Element> anchorScopeForAnchorName(const RenderBoxModelOb
     return nullptr;
 }
 
+enum class TopLayerStatus : uint8_t { Same, Lower, Higher };
```
I think it's `std::strong_ordering` / `std::partial_ordering`, no strong opinion if we should use it here, but maybe other folks do

anttijk committed:
[css-anchor-position-1] Anchored element should be in the same or higher top layer than the anchor
https://bugs.webkit.org/show_bug.cgi?id=295516
rdar://155216924

Reviewed by Tim Nguyen.

https://drafts.csswg.org/css-anchor-position-1/#acceptable-anchor-element

possible anchor is laid out strictly before positioned el, aka one of the following is true:
- positioned el is in a higher top layer than possible anchor
- Both elements are in the same top layer..

* LayoutTests/TestExpectations:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
(WebCore::Style::computeTopLayerStatus):

Canonical link: https://commits.webkit.org/297070@main
